proposals/0004.md

```diff
@@ -0,0 +1,153 @@
+* Title: *Radicle client URI scheme*
+* Champion: *@rudolfs, rudolfs@osins.org*
+* Conversation:
+  * [Radicle Community discussion][community]
+  * [Shareable identifier discussion][shareable]
+  * [Deep linking discussion][deeplink]
+* References:
+  * [List of URI schemes][urilist]
+
+# Motivation
+
+We have received multiple feature requests from our community regarding
+deep-linking in our client application. A common way to enable such
+functionality is to devise a URI scheme along with a custom protocol and
+implement a handler which listens for requests on the given protocol in a
+client application.
+
+Given that there might be multiple competing Radicle client implementations and
+to make sure that shared URIs are compatible with as many Radicle client
+implementations as possible we decided to have a specification for our URI
+scheme.
+
+
+# Explanation
+
+The main use case for a custom protocol is to be able to open a Radicle client
+application by clicking on a link on the web in an external browser. The
+Radicle client should be able to recognise what kind of resource is being
+requested and be able to render the request accordingly.
+
+Besides showing resources it might be valuable to perform certain actions on
+said resources. For example a user may be able to start tracking a project by
+clicking a link or initiate a fund transfer to a certain address.
+
+
+# Recommendation
+
+To make implementation easier we want to use existing URI parsing libraries,
+therefore our URI scheme should be parsable according to [RFC 3986][rfc3986].
+
+To make sure our URI scheme is extensible in the future and clients can provide
+backwards compatibility for older links we should have a version identifier in
+the scheme.
+
+The URI scheme to be used by Radicle client applications should adhere to the
+following scheme:
+
+    radicle://[<protocolVersion>/<context>/<entity>?action=<action>]
```

Comment:
Would it make sense to put the protocol version into the protocol name? E.g. now we have `radicle://...`, but when we introduce a breaking change, we switch to `radicle2://...`? How often do we expect the breaking changes to occur? I can think of these side effects:
- Shorter, clearer URIs
- Simpler parsers
- Introduction of `radicle2://...` requires registration of a new schema in the OS

Reply:
Hm, that would mean every time there's a change to the protocol, we have to create a new scheme. This means that with time there will be a lot of different links floating around on the web (`radicle://`, `radicle1://`, `radicle2://`, etc), which might be confusing to users. And as you point out, every protocol version change would require us to register a new protocol, which normally happens on app install. This means that newer versions of links would not processed by an older client at all.

I'm now leaning more towards what @kim [suggested](https://github.com/radicle-dev/radicle-decisions/pull/10#discussion_r592978118) — moving the version after the `<context>`, so each context can be versioned separately.

Reply:
> How often do we expect the breaking changes to occur?

Good question. One change that happened recently was the switch from one `Radicle ID` type to another in the `link` context:
  - old: `rad:git:hwd1yrey6bhkhk9g58fyhndo18ihj3aikr8p3c3f8byxir6saqegkdifoje`
  - new: `rad:git:hnrkkp3wifc691fw9j9sfed1adp5swfnu3npy`).

I don't expect such changes to happen frequently, most changes should be additive and not break backwards compatibility.

Reply:
Sounds good, thank you for the explanation :+1:

Reply:
nit: this is not technically what made the change breaking. Everything after `rad:` can change, but still be parsed unambiguously. Which is why I'm arguing to think about this in the same way as one would think about a RESTful API, only with a different scheme, and with some restrictions.